doc/rmcpping.8.in
\."#############################################################################
\."$Id: rmcpping.8.in,v 1.1 2004-05-13 17:32:57 chu11 Exp $
\."#############################################################################
\."  Copyright (C) 2003 The Regents of the University of California.
\."  Produced at Lawrence Livermore National Laboratory (cf, DISCLAIMER).
\."  Written by Albert Chu <chu11@llnl.gov>
\."  UCRL-CODE-155448
\."
\."  This file is part of Ipmiping, tools for pinging IPMI and RMCP compliant
\."  remote systems. For details, see http://www.llnl.gov/linux/.
\."
\."  Ipmiping is free software; you can redistribute it and/or modify it under
\."  the terms of the GNU General Public License as published by the Free
\."  Software Foundation; either version 2 of the License, or (at your option)
\."  any later version.
\."
\."  Ipmiping is distributed in the hope that it will be useful, but WITHOUT
\."  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
\."  FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
\."  for more details.
\."
\."  You should have received a copy of the GNU General Public License along
\."  with Ipmiping; if not, write to the Free Software Foundation, Inc.,
\."  59 Temple Place, Suite 330, Boston, MA  02111-1307  USA.
\."############################################################################
.TH RMCPING 8 "April 2004" "rmcpping @VERSION@" "System Manager Commands"
.SH NAME
rmcpping \- send RMCP Ping to network hosts
.SH SYNOPSIS
.B rmcpping [OPTIONS] destination
.SH DESCRIPTION
.B rmcpping
uses the RMCP Ping request datagram to elicit an RMCP Pong response from a
remote host.  The utility can be used to verify if a remote host supports
RMCP or IPMI.

.B rmcpping
will return 0 to the environment if it receives atleast 1 response
from the remote host.  Otherwise, it exits with a value of 1.
.SH OPTIONS
The following options are available 
.TP
.I "-h"
Output help menu.
.TP
.I "-V"
Output version.
.TP
.I "-c count"
Stop after sending 
.I count
ping packets.
.TP
.I "-i interval"
Wait 
.I interval 
seconds between sending each packet.  The default is to wait for one
second between each packet normally.
.TP
.I "-I interface address"
Set source address to specified interface address. Argument may
be numeric IP address or name of device.
.TP
.I "-t timeout"
Time to wait for a response, in seconds. Default is 5 seconds.
.TP
.I "-v"
Verbose output.
.if @WITH_DEBUG@ \{
.TP
.I "-d"
Turn on debugging.
\}
.SH "KNOWN ISSUES"
It has been observed that some remote BMCs can get "confused" and
delay packet responses if duplicate packets are sent in succession
very quickly.  For example, this could happen if the user repeatedly
executes "rmcpping -c 1 destination" very quickly.  There is no known
way to cleanly deal with a "confused" BMC.  The best option is to
simply wait "awhile".
.SH "ORIGIN"
Developed by Albert Chu <chu11@llnl.gov>.  Command and manpage
based off 
.B ping(8).
.SH "SEE ALSO"
ping(8), ipmiping(8)


NEWS
FreeIPMI 1.7.1 - XX/XX/XX
-------------------------

In 2020 the IPMI standard was frozen.  So a decision was made to try
and make the FreeIPMI 1.7.X line the last stable line in FreeIPMI.  A
number of decisions were made to remove legacy support and fix minor issues,
even at the cost of breaking backwards compatibility.

Updated Features:
-----------------

Globally
--------
o Internally allow hostnames to be up to 256 characters (previously only 64).

Libraries
---------
o libfreeipmi/sensor-read: Add error for privilege level insufficient.
o In libfreeipmi, support Intel S2600BPB OEM SEL interpretations.
o Significant refactoring of Intel OEM extensions.  Names of
  variables, macros, etc. may have changed.

Tools
-----
o In ipmi-sel, support Intel S2600BPB OEM SEL interpretations.
o In ipmi-config, support configuration of Maximum_Privilege_Cipher_Suite_Id_16.
o In ipmi-sensors, support new "altbridging" workaround for Lenovo motherboards.

Smaller Bug Fixes
-----------------

o In ipmi-config, drop support for
  Rmcpplus_Conf_Privilege:Maximum_Privilege_Cipher_Suite_Id_16.  It is
  not configurable.

o Globally, fix potential output sorting error with
  --consolidate-output.  Could alter output differently than before.

o Fix many minor output errors (i.e. typos, space alignment, lack of
  newline, etc.).  Could cause issues for those who script against
  output.

o In libfreeipmi, add missing fields for BMC Message Channel SDR
  record format.

Remove Large Amount of Legacy Support
-------------------------------------

Remove backwards compatibility, legacy output/options/fields/files, or
deprecated options.  All verified to have been gone for atleast 10
years.

For details of changes see NEWS.errata

Typo Fixes:
-----------

Fix large number of typo and minor output formattings.  They have the
potential to break against existing builds or scripts or configs that
relied on previous expected output.

See NEWS.errata for details

FreeIPMI 1.6.1 - 02/02/18
-------------------------
o Add IPv6 hostname support to FreeIPMI, all of FreeIPMI can now take
  IPv6 addresses as inputs to "host" parameters, options, or inputs.
o Support significant portions of IPMI IPv6 configuration in
  libfreeipmi.
o Add --no-session option in ipmi-raw.
o Add SDR cache options to ipmi-config.
o Legacy -f short option for --flush-cache and -Q short option for
  quiet-cache.  Backwards compatible for tools that supported it
  before.
o In ipmi-oem, support Gigabyte get-bmc-services and set-bmc-services.
o Various performance improvements:
  - Remove excessive calls to secure_memset to clear memory.
  - Remove excessive memsets and clears of data.
  - Remove unnecessary "double input checks".
  - Remove expensive input checks in libfreeipmi fiid library.
    Fallout from this may include FIID_ERR_FIELD_NOT_FOUND errors in
    different fiid functions.
  - Remove unnecessary input checks in libfreeipmi fiid library.
  - Add recent 'lookups' of fields in fiid library to internal cache.
o Various minor fixes/improvements
  - Update libfreeipmi core API to use poll() instead of select(), to
    avoid issues with applications with a high number of threads.

FreeIPMI 1.5.7 - 08/16/17
-------------------------
o In libipmimonitoring, fix several mem-leak corner cases.

FreeIPMI 1.5.6 - 07/05/17
-------------------------
o In libfreeipmi locate (used by ipmi-locate), use DMI firmware in
  sysfs if it exists.
o Minor mem-leak corner case fix in libfreeipmi.
o Minor documentation fixes.
o Minor error message clarifications.

FreeIPMI 1.5.5 - 11/02/16
-------------------------
o Fix invalid flag clear in libipmiconsole that can lead to a potential
  double close on a file descriptor.
o Support Supermicro H8SGL-F OEM sensors and events.

FreeIPMI 1.5.4 - 09/30/16
-------------------------
o Various changes/fixes in libipmiconsole
  - If user retrieves file descriptor from ipmiconsole_ctx_fd, user is
    required to close it.  ipmiconsole_ctx_destroy no longer closes
    it.  This is to avoid a potential double close which can be a
    problem for multithreaded applications.  This is a change in
    behavior, but we do not believe this will affect most applications
    since most users close the file descriptor under most scenarios
    anyways.
  - ipmiconsole_ctx_destroy() should now be called to free resources
    even if ipmiconsole_engine_teardown() has been called.  This has
    been done to create consistent behavior in the API and avoid a
    former segfault possibility.  This is a change in behavior, but we
    do not believe this will affect most applications since
    ipmiconsole_engine_teardown() is only called when an application
    is being shutdown.
  - The use of IPMICONSOLE_ENGINE_CLOSE_FD has been clarified in the
    header file.  Some of the prior text was unclear.  Behavior has
    not been changed.
o In ipmi-oem, support Intel get-bmc-services and set-bmc-services commands.
o In ipmi-oem, support Gigabyte get-nic-mode and set-nic-mode commands.
o Support Gigabyte MD90-FS0-ZB OEM SEL events.

FreeIPMI 1.5.3 - 08/01/16
-------------------------
o Update Fujitsu OEM extension product id support range.
o In ipmiconsole, support --debugfile option for more difficult
  debugging scenarios.
o In all sensor related tools, increase max number of record ids and
  sensor types that can be handled.
o Fix several invalid completion code errors/values.

FreeIPMI 1.5.2 - 05/24/16
-------------------------
o Update configuration file parsing code to handle values up to 1024
  characters in length.
o Set FD_CLOEXEC for ipmi driver device files opened within
  libfreeipmi.
o Support --read-fru, --write-fru, and --device-id option in
  bmc-device.
o Support --fru-file option in ipmi-fru.
o Various library updates in libfreeipmi to handle new features.

FreeIPMI 1.5.1 - 12/01/15
-------------------------
o Update Intel Node Manager support for 3.0 specification.
  - Update API and fiid templates in libfreeipmi.

rmcpping/rmcpping.c
/*****************************************************************************\
 *  $Id: rmcpping.c,v 1.55 2010-02-08 22:02:31 chu11 Exp $
 *****************************************************************************
 *  Copyright (C) 2007-2015 Lawrence Livermore National Security, LLC.
 *  Copyright (C) 2003-2007 The Regents of the University of California.
 *  Produced at Lawrence Livermore National Laboratory (cf, DISCLAIMER).
 *  Written by Albert Chu <chu11@llnl.gov>
 *  UCRL-CODE-155448
 *
 *  This file is part of Ipmiping, tools for pinging IPMI and RMCP compliant
 *  remote systems. For details, see https://savannah.gnu.org/projects/freeipmi/.
 *
 *  Ipmiping is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by the
 *  Free Software Foundation; either version 3 of the License, or (at your
 *  option) any later version.
 *
 *  Ipmiping is distributed in the hope that it will be useful, but
 *  WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 *  or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 *  for more details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with Ipmiping.  If not, see <http://www.gnu.org/licenses/>.
\*****************************************************************************/

#if HAVE_CONFIG_H
#include "config.h"
#endif /* HAVE_CONFIG_H */

#include <stdio.h>
#include <stdlib.h>
#if STDC_HEADERS
#include <string.h>
#endif /* STDC_HEADERS */
#if HAVE_UNISTD_H
#include <unistd.h>
#endif /* HAVE_UNISTD_H */
#include <assert.h>
#include <errno.h>

#include <freeipmi/freeipmi.h>

#include "freeipmi-portability.h"
#include "debug-util.h"
#include "ping-tool-common.h"

#define _supported(x)   ((x) ? "supported" : "not-supported")

int
createpacket (const char *destination,
              void *buf,
              unsigned int buflen,
              unsigned int sequence_number,
              int version,
              int debug)
{
  fiid_obj_t obj_rmcp_hdr = NULL;
  fiid_obj_t obj_rmcp_cmd = NULL;
  int len;

  assert (destination);
  assert (buf);

  if (!buflen)
    return (0);

  if (!(obj_rmcp_hdr = fiid_obj_create (tmpl_rmcp_hdr)))
    ipmi_ping_err_exit ("fiid_obj_create: %s", strerror (errno));
  if (!(obj_rmcp_cmd = fiid_obj_create (tmpl_cmd_asf_presence_ping)))
    ipmi_ping_err_exit ("fiid_obj_create: %s", strerror (errno));

  if (fill_rmcp_hdr_asf (obj_rmcp_hdr) < 0)
    ipmi_ping_err_exit ("fill_rmcp_hdr_asf: %s", strerror (errno));

  /* Avoid use of message tag number 0xFF.  Behavior of message tag 0xFF
   * is unpredictable.  See IPMI 1.5 Specification and DMTF RMCP
   * specification for details.
   */

  if (fill_cmd_asf_presence_ping (sequence_number % (RMCP_ASF_MESSAGE_TAG_MAX + 1),
                                  obj_rmcp_cmd) < 0)
    ipmi_ping_err_exit ("fill_cmd_asf_presence_ping: %s", strerror (errno));

  if ((len = assemble_rmcp_pkt (obj_rmcp_hdr,
                                obj_rmcp_cmd,
                                buf,
                                buflen,
                                IPMI_INTERFACE_FLAGS_DEFAULT)) < 0)
    ipmi_ping_err_exit ("assemble_rmcp_pkt: %s", strerror (errno));

  if (debug)
    {
      char hdrbuf[DEBUG_UTIL_HDR_BUFLEN + 1];

      memset (hdrbuf, '\0', DEBUG_UTIL_HDR_BUFLEN + 1);

      debug_hdr_str (DEBUG_UTIL_TYPE_NONE,
                     DEBUG_UTIL_DIRECTION_NONE,
                     DEBUG_UTIL_FLAGS_DEFAULT,
                     DEBUG_UTIL_RMCPPING_STR,
                     hdrbuf,
                     DEBUG_UTIL_HDR_BUFLEN);

      if (ipmi_dump_rmcp_packet (STDERR_FILENO,
                                 destination,
                                 hdrbuf,
                                 NULL,
                                 buf,
                                 len,
                                 tmpl_cmd_asf_presence_ping) < 0)
        ipmi_ping_err_exit ("ipmi_dump_rmcp_packet: %s", strerror (errno));
    }

  fiid_obj_destroy (obj_rmcp_hdr);
  fiid_obj_destroy (obj_rmcp_cmd);
  return (len);
}

int
parsepacket (const char *destination,
             const void *buf,
             unsigned int buflen,
             const char *from,
             unsigned int sequence_number,
             int verbose,
             int version,
             int debug)
{
  fiid_obj_t obj_rmcp_hdr = NULL;
  fiid_obj_t obj_rmcp_cmd = NULL;
  uint8_t message_type, message_tag;
  uint64_t val;
  int ret, rv = -1;

  assert (destination);
  assert (buf);
  assert (from);

  if (!buflen)
    return (0);

  if (!(obj_rmcp_hdr = fiid_obj_create (tmpl_rmcp_hdr)))
    ipmi_ping_err_exit ("fiid_obj_create: %s", strerror (errno));
  if (!(obj_rmcp_cmd = fiid_obj_create (tmpl_cmd_asf_presence_pong)))
    ipmi_ping_err_exit ("fiid_obj_create: %s", strerror (errno));

  if (debug)
    {
      char hdrbuf[DEBUG_UTIL_HDR_BUFLEN + 1];

      memset (hdrbuf, '\0', DEBUG_UTIL_HDR_BUFLEN + 1);

      debug_hdr_str (DEBUG_UTIL_TYPE_NONE,
                     DEBUG_UTIL_DIRECTION_NONE,
                     DEBUG_UTIL_FLAGS_DEFAULT,
                     DEBUG_UTIL_RMCPPONG_STR,
                     hdrbuf,
                     DEBUG_UTIL_HDR_BUFLEN);

      if (ipmi_dump_rmcp_packet (STDERR_FILENO,
                                 destination,
                                 hdrbuf,
                                 NULL,
                                 buf,
                                 buflen,
                                 tmpl_cmd_asf_presence_pong) < 0)
        ipmi_ping_err_exit ("ipmi_dump_rmcp_packet: %s", strerror (errno));
    }

  if ((ret = unassemble_rmcp_pkt (buf,
                                  buflen,
                                  obj_rmcp_hdr,
                                  obj_rmcp_cmd,
                                  IPMI_INTERFACE_FLAGS_DEFAULT)) < 0)
    ipmi_ping_err_exit ("unassemble_rmcp_pkt: %s", strerror (errno));

  if (!ret)
    {
      rv = 0;
      goto cleanup;
    }

  if (FIID_OBJ_GET (obj_rmcp_cmd,
                    "message_type",
                    &val) < 0)
    ipmi_ping_err_exit ("fiid_obj_get: 'message_type': %s",
                        fiid_obj_errormsg (obj_rmcp_cmd));
  message_type = val;

  if (message_type != RMCP_ASF_MESSAGE_TYPE_PRESENCE_PONG)
    {
      rv = 0;
      goto cleanup;
    }

  if (FIID_OBJ_GET (obj_rmcp_cmd,
                    "message_tag",
                    &val) < 0)
    ipmi_ping_err_exit ("fiid_obj_get: 'message_tag': %s",
                        fiid_obj_errormsg (obj_rmcp_cmd));
  message_tag = val;

  if (message_tag != (sequence_number % (RMCP_ASF_MESSAGE_TAG_MAX + 1)))
    {
      rv = 0;
      goto cleanup;
    }

  printf ("pong received from %s: message_tag=%u", from, message_tag);
  if (verbose)
    {
      uint8_t ipmi_supported;

      if (FIID_OBJ_GET (obj_rmcp_cmd,
                        "supported_entities.ipmi_supported",
                        &val) < 0)
        ipmi_ping_err_exit ("fiid_obj_get: 'supported_entities.ipmi_supported': %s",
                            fiid_obj_errormsg (obj_rmcp_cmd));
      ipmi_supported = val;

      printf (", ipmi %s", _supported (ipmi_supported));
    }
  printf ("\n");

  rv = 1;
 cleanup:
  fiid_obj_destroy (obj_rmcp_hdr);
  fiid_obj_destroy (obj_rmcp_cmd);
  return (rv);
}

void
latepacket (unsigned int sequence_number)
{
  printf ("pong timed out: message_tag=%u\n", sequence_number % (RMCP_ASF_MESSAGE_TAG_MAX + 1));
}

int
endresult (const char *progname,
           const char *destination,
           unsigned int sent_count,
           unsigned int recv_count)
{
  double percent = 0;

  assert (progname);
  assert (destination);

  if (sent_count > 0)
    percent = ((double)(sent_count - recv_count)/sent_count)*100;

  printf ("--- %s %s statistics ---\n", progname, destination);
  printf ("%u pings transmitted, %u pongs received in time, "
          "%2.1f%% packet loss\n",
          sent_count, recv_count, percent);

  return ((recv_count > 0) ? 0 : 1);
}

int
main (int argc, char **argv)
{
  ipmi_ping_setup (argc, argv, 0, RMCP_ASF_MESSAGE_TAG_MAX, "hVc:i:I:t:vs:d");
  ipmi_ping_loop (createpacket, parsepacket, latepacket, endresult);
  exit (EXIT_FAILURE);                    /* NOT REACHED */
}
